### PR TITLE
Restructure the "Creating an XML sitemap tutorial"

### PR DESCRIPTION
**Pull request by sofietoft**
Tasks

* [x] Rewrite steps into ordered lists (progress)
* [x] Fix wrong code samples

### CODE REVIEW

File: umbraco-cms/tutorials/creating-an-xml-site-map.md [hunk 1]
```diff
@@ -208,146 +232,169 @@ If we create a helper called `RenderSiteMapUrlEntriesForChildren` that accepts a
...
+
+* Not **hidden**,
+* Not based on an **excluded Document Type**, and
+* Located within the bounds of the defined **depth**.
+
+## The finished sitemap template
+
+**Add this to an expandable.**
 
 ```csharp
```
Comment (eshanrnh) [inline umbraco-cms/tutorials/creating-an-xml-site-map.md:358]: This code block needs to be updated. It contains the old code.

File: umbraco-cms/tutorials/creating-an-xml-site-map.md [hunk 2]
```diff
@@ -208,146 +232,169 @@ If we create a helper called `RenderSiteMapUrlEntriesForChildren` that accepts a
...
-        @RenderSiteMapUrlEntry(page)
-        if (page.Level < maxSiteMapDepth && page.Children.Any(f => !excludedDocumentTypes.Contains(f.ContentType.Alias) && !f.Value<bool>("hideFromXmlSiteMap")))
+        // Filter the query based on the excludedDocumentTypes value
+        foreach (var page in parentPage.Children.Where(f => !excludedDocumentTypes.Contains(f.ContentType.Alias) && !f.Value<bool>("hideFromXmlSiteMap")))
         {
-            @RenderSiteMapUrlEntriesForChildren(page, maxSiteMapDepth, excludedDocumentTypes)
+            RenderSiteMapUrlEntry(page);
+            if (page.Level < maxSiteMapDepth && page.Children.Any(f => !f.Value<bool>("hideFromXmlSiteMap")))
+            {
+                RenderSiteMapUrlEntriesForChildren(page, maxSiteMapDepth, excludedDocumentTypes);
```
Comment (eshanrnh) [inline umbraco-cms/tutorials/creating-an-xml-site-map.md:340]: This line throws an error:

image

File: umbraco-cms/tutorials/creating-an-xml-site-map.md [hunk 3]
```diff
@@ -33,171 +34,194 @@ See [Sitemaps XML format](https://www.sitemaps.org/protocol.html) for the XML sc
...
+    ```
 
-So hopefully you can see the problem here, how deep do we go? How do we handle the repetition forever...
+Do you see the potential problem here? How deep do we go and how do we handle the repetition?
 
-... well we can use recursion - we can create a further razor helper that 'calls itself' \[insert inception reference here]...
+One option is to use recursion. We can create a Razor helper that 'calls itself'.
 
-#### Recursive Helper
+{% hint style="warning" %}
```
Comment (eshanrnh) [inline umbraco-cms/tutorials/creating-an-xml-site-map.md:215]: I understand we are trying to show that this is one way of doing it... but I'm not sure if we really want users to do it and remove the code. There is a possibility they might just skip through paras/info and only look at the code. Maybe just skip that part and show what is supposed to be done since it's a tutorial and one might follow all the steps mentioned here 🙈

**Review by eshanrnh — COMMENTED** [review]
Overall looks quite good, @sofietoft 💪 I have tested out this PR and left some comments.

General Comments:
1) I think it will be a good idea to add screenshots/code block/expandables to show how the template looks like after updating each code block as it can be a little confusing as to where the code should go.
2) Update the old screenshots as document types now have **Add tab** button and frontend screenshots still shows **2018**

File: umbraco-cms/tutorials/creating-an-xml-site-map.md [hunk 4]
```diff
@@ -33,171 +34,194 @@ See [Sitemaps XML format](https://www.sitemaps.org/protocol.html) for the XML sc
...
+    <urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" 
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
+            xsi:schemalocation="http://www.google.com/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd" 
+            xmlns:image="http://www.google.com/schemas/sitemap-image/1.1">
+        // Insert sitemap content here.
+    </urlset>
+    ```
 
-We will retrieve each page in the site as **IPublishedContent**, and read in the SearchEngineChangeFrequency(recursively), SearchEngineRelativePriority, Url, when it was last modified etc...
+6. Save the template.
```
[the remark this answers is not in the document]
Comment (sofietoft) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:123]: I generally don't think we should use screenshots for code samples, as that takes away the option to copy the code.

I'll see about adding a preview of the template after each step in the tutorial.

File: umbraco-cms/tutorials/creating-an-xml-site-map.md [hunk 5]
```diff
@@ -33,171 +34,194 @@ See [Sitemaps XML format](https://www.sitemaps.org/protocol.html) for the XML sc
...
-        var priority = node.HasValue("searchEngineRelativePriority") ? node.Value<string>("searchEngineRelativePriority") : "0.5";
+The sitemap should start at the homepage at the root of the site. Since our XmlSiteMap page is created as a subpage page to the root, we can use the `Root()` helper to define the starting point as `IPublishedContent`.
 
-        <url>
-            <loc>@node.Url(mode: UrlMode.Absolute)</loc>
-            <lastmod>@(string.Format("{0:s}+00:00", node.UpdateDate))</lastmod>
-            <changefreq>@changeFreq</changefreq>
-            <priority>@priority</priority>
-        </url>
+1. Add `IPublishedContent siteHomePage = Model.Root();` within the first set of curly brackets in the template.
```
[the remark this answers is not in the document]
Comment (sofietoft) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:129]: Wouldn't that take away some consistency? 🤔

File: umbraco-cms/tutorials/creating-an-xml-site-map.md [hunk 6]
```diff
@@ -33,171 +34,194 @@ See [Sitemaps XML format](https://www.sitemaps.org/protocol.html) for the XML sc
...
 
-    }
+### Rendering a sitemap entry
 
-}
-```
+We will retrieve each page in the site as **IPublishedContent** and read in the properties we added in step 3: `SearchEngineChangeFrequency` and `SearchEngineRelativePriority`. We will also read the URL of the page as well as when it was last modified.
```
[the remark this answers is not in the document]
Comment (sofietoft) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:133]: Ah! It should actually be Step 2 of the tutorial. But it's a good point about how that can be confusing.
I think adding your suggestion though, would make the sentence longer than permitted 😅 

How about this instead:
> We will retrieve each page in the site as **IPublishedContent** and read in the `SearchEngineChangeFrequency` and `SearchEngineRelativePriority` properties.

File: umbraco-cms/tutorials/creating-an-xml-site-map.md [hunk 7]
```diff
@@ -33,171 +34,194 @@ See [Sitemaps XML format](https://www.sitemaps.org/protocol.html) for the XML sc
...
+Before moving on, remove any code pieces related to the two steps directly above this box.
+The code in those steps was to showcase what not to do.
+{% endhint %}
 
-If we create a helper called `RenderSiteMapUrlEntriesForChildren` that accepts a 'Parent Page' parameter as the starting point. Then we can find the children of this Parent Page, write out their Site Map Entry, and then call this same method again from itself - recursion!
+We will add a `RenderSiteMapUrlEntriesForChildren` helper which accepts a 'Parent Page' parameter as the starting point. Then we will find the children of this Parent Page and write out their sitemap entry. Finally, we will call this same method again from itself.
 
-```csharp
-@{
+1. Add the following code snippet below the `RenderSiteMapUrlEntry` helper:
```
[the remark this answers is not in the document]
Comment (sofietoft) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:222]: How about
" ... and before the closing curly bracket" , as I've called them that throughout the guide? 🤔

File: umbraco-cms/tutorials/creating-an-xml-site-map.md [hunk 8]
```diff
@@ -208,146 +232,169 @@ If we create a helper called `RenderSiteMapUrlEntriesForChildren` that accepts a
...
+                RenderSiteMapUrlEntriesForChildren(page);
+            }
         }
     }
-}
-```
+    ```
 
-Now revisit a page in the content tree, and tick the Hide from SiteMap option, if all has gone well, it will disappear from the XmlSitemap page!
+Revisit a page in the Content tree, and check the HideFromXmlSitemap option. This page will now be excluded from the sitemap.
```
[the remark this answers is not in the document]
Comment (sofietoft) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:280]: I am not entirely sure what a screenshot would be necessary here?
It is not a necessary step, just a way to verify that hiding a page from the sitemap works.

File: umbraco-cms/tutorials/creating-an-xml-site-map.md [hunk 9]
```diff
@@ -208,146 +232,169 @@ If we create a helper called `RenderSiteMapUrlEntriesForChildren` that accepts a
...
 
-**Depth**
+Another way to control which and how many pages are shown in the sitemap is to filter by **depth**. We will provide the helper with a number that defines how deep into the Content tree the sitemap should look.
```
[the remark this answers is not in the document]
Comment (sofietoft) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:282]: Ah! This part is not replacing the other or acting as an alternative. You can do both, and that's also what this tutorial does.

So perhaps I'll rephrase it to something like
"It is also possible to filter the sitemap entries based on the pages location in the Content tree."

What do you think about that?

File: umbraco-cms/tutorials/creating-an-xml-site-map.md (hunk 1 above)
Comment (sofietoft) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:358]: I am not entirely sure I understand what you mean about "the old code"?
Could you perhaps be a bit more specific as to where this code block need to be updated?

File: umbraco-cms/tutorials/creating-an-xml-site-map.md [hunk 10]
```diff
@@ -356,48 +403,46 @@ void RenderSiteMapUrlEntriesForChildren(IPublishedContent parentPage)
...
+## Going further
 
-Finally let search engines know the url for your sitemap by updating your robots.txt file accordingly:
+Once you have added a sitemap to your site it is recommended that you also reference it in your `robots.txt` file.
 
-```
-Sitemap: https://www.yourlovelysite.com/xmlsitemap
-User-agent: *
-```
+1. Locate and open the `robots.txt` file in your preferred IDE.
```
[the remark this answers is not in the document]
Comment (sofietoft) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:429]: This is a file that you would usually have in a project to configure how search engine bots crawl your website.
I didn't add a step to create the file, as I assume this is something that most people would already have in their project 💪 Also, creating one, would go a bit beyond the scope of this project - which I guess this part of the tutorial already does 😅

File: umbraco-cms/tutorials/creating-an-xml-site-map.md (hunk 2 above)
Comment (sofietoft) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:340]: Yes, the parameters need to be taken out - I'll update the sample 💪

**Review by sofietoft — COMMENTED** [review]
Thanks for the review @eshanrnh 💪 
I've accepted most suggestions - please also see my comments!

File: umbraco-cms/tutorials/creating-an-xml-site-map.md [hunk 11]
```diff
@@ -33,171 +34,194 @@ See [Sitemaps XML format](https://www.sitemaps.org/protocol.html) for the XML sc
...
+Create and configure the Document Type Composition by following these steps:
 
-At this point your composition should look similar to this:&#x20;
+1. Navigate to the **Settings** section in the Umbraco backoffice.
+2. Create a new **Composition** under the Document Types folder.
+3. Name the new Document Type **XmlSiteMapSettings**.
+4. Add the following properties:
+    a. Slider named **Search Engine Relative Priority** (searchEngineRelativePriority): MinValue: 0.1, MaxValue: 1, Step Increments 0.1, InitialValue 0.5.
+    b. Dropdown named **Search Engine Change Frequency** (searchEngineChangeFrequency): Always, hourly, daily, weekly, monthly, yearly, and never.
+    c. Checkbox named **Hide From Xml Sitemap** (hideFromXmlSitemap).
```
[the remark this answers is not in the document]
Comment (sofietoft) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:85]: Which Data Type did you use instead then? 😅 

I think Checkbox was the old name for "Toggle" 🤔 I'll investigate and make sure to update the name!

File: umbraco-cms/tutorials/creating-an-xml-site-map.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (eshanrnh) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:123]: Yes, a code block/template preview is completely fine as far as we provide the option to verify the code in the template looks the same

File: umbraco-cms/tutorials/creating-an-xml-site-map.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (eshanrnh) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:129]: Yes it will but currently it looks like an incomplete list.

File: umbraco-cms/tutorials/creating-an-xml-site-map.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (eshanrnh) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:222]: sounds good 💪

File: umbraco-cms/tutorials/creating-an-xml-site-map.md (hunk 1 above)
Comment (eshanrnh) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:358]: I was confused with the last part of the code as it threw an error for me. Now that I run the last part of the code from the sample below it works fine 💪

File: umbraco-cms/tutorials/creating-an-xml-site-map.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (eshanrnh) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:429]: Ah ok, then it's fine. I didn't know this piece of information so was looking around for the file 🙈

File: umbraco-cms/tutorials/creating-an-xml-site-map.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (eshanrnh) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:85]: I used Toggle data type.

File: umbraco-cms/tutorials/creating-an-xml-site-map.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (eshanrnh) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:280]: If it's not a necessary step then its fine to leave as is 💪

File: umbraco-cms/tutorials/creating-an-xml-site-map.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (eshanrnh) [reply umbraco-cms/tutorials/creating-an-xml-site-map.md:282]: This line reads better 🙌

**Comment by sofietoft** [comment]
I think I've covered everything from your review @eshanrnh ! Please take another look to confirm 💪 

I'll add the code snippets in expendables directly to GitBook - there is a couple of other changes that I need to make there as well.

**Comment by sofietoft** [comment]
Added code snippets to expandables 💪 @eshanrnh

**Review by eshanrnh — APPROVED** [review]
Looks good to me, @sofietoft  🙌
